### DISCUSSION TITLE
I am trying to use the read and write of a websocket in different threads.

### DISCUSSION THREAD

**Discussion by lz1998** [tokio-rs/axum#282] [Q&A] [answered]
https://github.com/tokio-rs/axum/blob/main/examples/websockets/src/main.rs

```rust
use std::sync::Arc;
use tokio::sync::Mutex;
async fn handle_socket(mut socket: WebSocket) {
    let socket=Arc::new(Mutex::new(socket));
    tokio::spawn(async move {
        loop {
            if let Some(msg) = socket.lock().recv().await {
                if let Ok(msg) = msg {
                    println!("Client says: {:?}", msg);
                } else {
                    println!("client disconnected");
                    return;
                }
            }
        }
    });

    loop {
        if socket
            .lock()
            .send(Message::Text(String::from("Hi!")))
            .await
            .is_err()
        {
            println!("client disconnected");
            return;
        }
        tokio::time::sleep(std::time::Duration::from_secs(3)).await;
    }
}
```

**Comment by davidpdrsn** [comment, marked as answer]
Since WebSocket implements Stream and Sink you can use [split](https://docs.rs/futures/0.3.16/futures/prelude/stream/trait.StreamExt.html#method.split) to get a read and write half. Those can be moved to different threads.

It's also shown here https://github.com/tokio-rs/axum/blob/main/examples/chat/src/main.rs

**Reply by davidpdrsn** [reply]
I've made an issue to document this as it's not very obvious https://github.com/tokio-rs/axum/issues/283

**Reply by lz1998** [reply]
![image](https://user-images.githubusercontent.com/9082086/131223306-f68075e2-ceed-4422-942e-211cd361fd8a.png)
Is there any examples?

**Reply by davidpdrsn** [reply]
Yes. The one I linked to.

**Reply by lz1998** [reply]
Sorry, I have unmarked this answer by mistake
